src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -708,19 +708,21 @@ private void Initialize(bool multipleServices)
 
                 _mainCallback = new ServiceMainCallback(this.ServiceMainCallback);
                 _commandCallbackEx = new ServiceControlCallbackEx(this.ServiceCommandCallbackEx);
-                _handleName = Marshal.StringToHGlobalUni(this.ServiceName);
 
                 _initialized = true;
             }
         }
 
+        // Make sure that the name field is freed after use. We allocate a new string to avoid holding one central handle,
```

Comment:
Nit: It is odd contract to return a partially marshaled structure like this out of a method: The delegate is not marshaled vs. the name is marshaled. 

I would be nice to:
- either make the `SERVICE_TABLE_ENTRY` structure fully managed: change `name` field to string (with annotations to marshal it as Unicode string), depend on `Marshal.StructureToPtr` to marshal the whole thing and `Marshal.DestroyStructure` to destroy it.
- or make the `SERVICE_TABLE_ENTRY` structure fully unmanaged: change `callback` field to IntPtr, and marshal the delegate manually using `Marshal.GetFunctionPointerForDelegate`.

We generally prefer the manual marshaling because of it has better performance and it makes it easy to see what is going on.

Reply:
I was thinking of doing the former, but I wasn't sure if how 'thin' the structs should be (i.e. whether it should hold "Handles" that we manually marshal, or managed objects that gets handled automatically by the marshaller).

I can change that.